Uncontaminated Sample Packing

djsaunde opened the pull request:
This PR adds sample packing support. It uses TRL's SFTConfig `packing=True` and `padding_free=True` args to pack the sequences, and we compute `packed_seq_lengths` metadata and thread it through the model forward pass. This metadata is used to create block causal masks for SDPA and xformers attention, and is passed to the flash attention varlen API which handles the block causal masking itself under the hood (we need to do this ourselves because of our custom forward pass, whereas TRL handles the sequence length metadata internally in their trainer).

I added a few unit tests. I also wrote a quick bash script for smoke testing some common model architectures: [gist](https://gist.github.com/djsaunde/080643f8657dcee07ddfc89ae86148fa), which runs.

Below is a comparison of short `unsloth/qwen2.5-0.5b` training runs. The losses don't match because we're seeing more / different samples on each step. But the scale and trend match, which is the important bit.

<img width="2138" height="665" alt="image" src="https://github.com/user-attachments/assets/b44df873-d236-4aec-ae21-aaa06065e7fc" />

Commands:

No sample packing:

```bash
python unsloth-cli.py --model_name unsloth/qwen2.5-0.5b --dataset yahma/alpaca-cleaned --per_device_train_batch_size 8 --max_steps 50 --max_seq_length 2048
```

Sample packing:

```bash
python unsloth-cli.py --model_name unsloth/qwen2.5-0.5b --dataset yahma/alpaca-cleaned --per_device_train_batch_size 1 --max_steps 50 --max_seq_length 2048 --sample_packing
```

Note that we use `--per_device_train_batch_size 1` in the latter case since we are packing multiple examples into a single `[1, max_seq_length]` tensor.

The benefit of this approach is that we're able to discard a lot of zero padding, and therefore get higher token/s training throughput. The below plot shows that we're able to get through our dataset ~20% faster. These gains depend on the dataset and configured `--max_seq_length`; if we increase this we generally get better packing efficiency => higher throughput.

<img width="723" height="321" alt="image" src="https://github.com/user-attachments/assets/de7b71df-caf1-420c-8643-336efff71864" />

I manually tested on SDPA and flash attention, but I still need to test xformers attention since I couldn't get it to build for blackwell.

TODO
- [x] test xformers attention

Datta0 commented on unsloth/kernels/rope_embedding.py:
```diff
@@ -145,20 +230,123 @@ def backward(ctx, dY):
+    else:
+        Q_out = Fast_RoPE_Embedding.apply(Q.transpose(1, 2).contiguous(), cos, sin).transpose(1, 2)
+        K_out = Fast_RoPE_Embedding.apply(K.transpose(1, 2).contiguous(), cos, sin).transpose(1, 2)
     torch_device_stream(Q.device).synchronize()
```
On this, this was added cuz multi GPU forward pass was returning garbage without this. Can you check if this is needed for your kernels as well? Or can we skip doing this if possible?

Also I remember benchmarking with and without this and saw a very minute performance diff on single GPU. If there's a significant hit, we can do 
```
if DEVICE_COUNT >1:
        torch_device_stream(Q.device).synchronize()
```

djsaunde replied:
hm this is currently block as DDP seems to be broken

djsaunde replied:
fixed DDP and added this check, which improves single GPU perf slightly (I wasn't able to repro the garbage outputs, but I think I understand why they can happen)

djsaunde commented:
I added some utils and updated the CLI to work OOTB with DDP. Just use `torchrun --nproc_per_node=N` or `accelerate launch` on a multi-GPU machine and it should just work.

These utils should be reusable in our notebooks / scripts too!

PS: DDP working relies on removing the `@torch.compile` decorator from `unsloth_zoo/patch_torch_functions.py::cross_entropy` as it results in a double compile somehow. I think @danielhanchen is fixing this.

djsaunde commented:
Closing in favor of #3566.
